core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +321,104 @@ private void balance() {
             balanceByWeights();
         }
 
+        /**
+         * Makes a decision about moving a single shard to a different node to form a more
+         * optimally balanced cluster.  This method is invoked from the cluster allocation
+         * explain API only.
+         */
+        private RebalanceDecision decideRebalance(final ShardRouting shard) {
+            if (shard.started() == false) {
+                // cannot rebalance a shard that isn't started
+                return RebalanceDecision.NOT_TAKEN;
+            }
+
+            Decision canRebalance = allocation.deciders().canRebalance(shard, allocation);
+            if (canRebalance.type() != Type.YES) {
+                // pick the first NO rebalance decision and use its explanation for the final explanation
+                String explanation = null;
+                Type explanationCause = null;
+                for (Decision subDecision : canRebalance.getDecisions()) {
+                    if ((subDecision.type() == Type.NO && (explanation == null || explanationCause == Type.THROTTLE))
+                            || (subDecision.type() == Type.THROTTLE && explanation == null)) {
+                        explanation = subDecision.getExplanation();
+                        explanationCause = subDecision.type();
```

Comment:
Yes, but if our first decision is THROTTLE, we wait to see if there is a NO decision as that is stronger.  However, once we have an explanation based on a NO decision, we can break, so I will add that check.